x-pack/elastic-agent/pkg/agent/application/upgrade/service.go

```diff
@@ -0,0 +1,89 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build !darwin
+// +build !windows
+
+package upgrade
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/coreos/go-systemd/v22/dbus"
+
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/application/paths"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/install"
+)
+
+const (
+	// delay after agent restart is performed to allow agent to tear down all the processes
+	// important mainly for windows, as it prevents removing files which are in use
+	afterRestartDelay = 2 * time.Second
+)
+
+// Init initializes os dependent properties.
+func (ch *CrashChecker) Init(ctx context.Context) error {
+	dbusConn, err := dbus.New()
```

Comment:
I love the use of DBUS, but I think this is specific to systemd. We need to also support sysv and upstart. I think the service module I used for `install/uninstall` has the ability to get information about a service in a generalized way. I don't know if it will give PID, but if so that would ensure that it works on all linux init systems.

Reply:
yes i was hoping for that as well. but it only provides a status meaning Running, Stopped ...
i will check how i can provide PIDers for other service managers